### ISSUE TITLE
Question: How is table updated in this case

### ISSUE THREAD

**Issue by malx122** [Starcounter/Starcounter#2524]
I make a simplified example:
```
[Database]
public class A
{
}
[Database]
public class B: A
{
}

[Database]
public class C
{
     public A MyProperty;
}
```
Lets say I create a number of C, B and A and store them in DB. Some of the C will have MyProperty pointing at objects of type A and some will have MyProperty pointing to some of type B. Since B is of type A there will be no problem.

But lets say I realize that I actually want to have them just point at objects of type B, so I change class C:
```
[Database]
public class C
{
     public B MyProperty;
}
```
What happen with the data already in the DB when "upgrading"?
1. All current instances of C will get a MyProperty pointing at null?
2. All instances of C that had a MyProperty of type B will keep the reference, the ones with a Myproperty of type A will be null?
3. Failure?

**Comment by k-rus** [comment]
Thank you for your question!
Exactly this case will work fine. Due to bug #1310 the difference between old and new `C` will not be recognized and since all presented objects, i.e., belonging to `A`, can be casted to `B`, you will not meet the problems. However it might be a problem in other situations.
Currently recommended approach for fixing such upgrade of a schema is

1. Add new property without removing the old one
2. Copy the data from the old property to the new property with correct casting
3. Remove the old property

If you want that the new property as the old one you need to repeat the steps.

**Comment by malx122** [comment]
No, if you look above all "B" can be casted to "A", not the opposite. And in my Db, "MyProperty" has both A and B:s in it. Thats why I had those 3 different outcome in my question.

**Comment by k-rus** [comment]
> No, if you look above all "B" can be casted to "A", not the opposite.

You are right. In this case the behavior is not defined. I guess it should raise a runtime error, since it cannot be casted from `A` to `B`. 
The upgrade script will be rewritten in next version of Starcounter, so now it is necessary to do the work around described above.
